experimental/personnames/src/api.rs

```diff
@@ -0,0 +1,161 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use std::mem::Discriminant;
+
+use icu_locid::Locale;
+
+/// Trait for providing person name data.
+pub trait PersonName {
+    /// Return the name locale of person name.
+    fn name_locale(&self) -> Option<&Locale>;
+
+    /// return the preferred order of person name.
+    fn preferred_order(&self) -> Option<&PreferredOrder>;
+
+    /// return the value of the given field name, it *must* match the name field.
+    fn get(&self, field: &NameField) -> Option<&str>;
+
+    /// Return all available name field.
+    fn available_name_fields(&self) -> Vec<&NameField>;
+
+    /// Returns true if the provided field name is available.
+    fn has_name_field(&self, lookup_name_field: Discriminant<NameField>) -> bool;
+
+    /// Returns true if person have the name field matching the type and modifier.
+    fn has_name_field_with_modifier(&self, lookup_name_field: &NameField) -> bool;
+}
+
+///
+/// Error handling for the person name formatter.
+///
+#[derive(Debug)]
+pub enum PersonNamesFormatterError {
+    ParseError(String),
+    InvalidPersonName,
+}
+
+/// Field Modifiers.
+///
+/// https://www.unicode.org/reports/tr35/tr35-personNames.html#modifiers
+#[repr(u32)]
+pub enum FieldModifier {
+    Informal = 1 << 0,
+    Prefix = 1 << 1,
+    Core = 1 << 2,
+    AllCaps = 1 << 3,
+    InitialCap = 1 << 4,
+    Initial = 1 << 5,
+    Monogram = 1 << 6,
+}
+
+/// Field Modifiers Mask. (must be the same as FieldModifier repr)
+pub type FieldModifierMask = u32;
+
+/// List all incompatible combination of field modifier.
+const MUTUALLY_EXCLUSIVE_FIELD_MODIFIERS: [FieldModifierMask; 3] = [
+    FieldModifier::Prefix as FieldModifierMask | FieldModifier::Core as FieldModifierMask,
+    FieldModifier::AllCaps as FieldModifierMask | FieldModifier::InitialCap as FieldModifierMask,
+    FieldModifier::Initial as FieldModifierMask | FieldModifier::Monogram as FieldModifierMask,
+];
+
+/// Returns true if the field modifier is valid.
+pub fn field_modifier_is_valid(mask: FieldModifierMask) -> bool {
+    MUTUALLY_EXCLUSIVE_FIELD_MODIFIERS
+        .into_iter()
+        .all(|field| mask & field != field)
+}
+
+/// Name Fields defined by Unicode specifications.
+///
+/// https://www.unicode.org/reports/tr35/tr35-personNames.html#fields
+#[derive(Ord, Eq, PartialEq, PartialOrd)]
+pub enum NameField {
```

Comment:
In ICU4J, NameField is a plain enum of Title/Given/Given2/…, and the masks are given as a separate standalone argument to functions like `PersonName::get`.

Also, can we just make the empty FieldModifierMask be the default value instead of wrapping it in an Option?

```
pub struct NameField {
    pub kind: NameFieldKind,
    pub modifiers: FieldModifierMask
}
```